### PR TITLE
feature(@desktop/chat): implement search on sqlcipher (status-go side)

### PR DESCRIPTION
**Pull request by saledjenic**
Current code adapted to handle future changes on message search (like searching message in multiple
channels). 

Memory leak which was happening in qml assigning (copying) MessageItem to qml variable messageItem
(where that qml variable messageItem was never deleted) is fixed.

Searching messages by some term for a specific channel is added on the side of status-go and an
appropriate part on the side of nim is developed accordingly.

Fixes: #2912

### CODE REVIEW

File: src/app/chat/views/messages.nim [hunk 1]
```diff
@@ -496,3 +479,46 @@ QtObject:
...
+    if (chatId.len == 0):
+      error "search messages response either doesn't contain chatId or it is empty"
+      return
+
+    let messagesObj = if(responseObj.contains("messages")): responseObj{"messages"} else: newJObject()
+    if (messagesObj.kind != JObject):
+      error "search messages response either doesn't contain messages object or it is empty"
+      return
+
+    let (cursor, messages) = libstatus_chat.parseChatMessagesResponse(messagesObj)
```
Comment (jrainville) [inline src/app/chat/views/messages.nim:506]: I know using `libstatus` directly in a view is not recommended. Maybe we could expose `parseChatMessagesResponse` in `status/chat.nim` instead

**Review by jrainville — COMMENTED** [review]
Looks good. Nice job.

I just put one small comment. Also, I'll let @michaelsbradleyjr or @emizzle check if the thread work is kosher to make sure no leak can happen

**Comment by saledjenic** [comment]
You may notice that I intentionally did async things in status/chat.nim, the reason why I did that is that I don't want to mix view only things and some logical stuff. I noticed that many status lib procedures are called directly from a view, what is not good and hard to maintain. Every view should only know what is important for displaying, nothing else, entire logic, including remote api calls, should be done in backend controllers (what is in our case code put in status folder) and those controllers via signal/slot mechanism should notify views about the changes. That's why I introduced new type "SlotArg", just to easier pass which method on which instance should be called once the job is finished (this is also not a perfect, but should be kind of a start).

Looking at our views, they are far from an ideal state, but I would like if you agree to start making a line what is a job for views, what for controllers. And step by step in few months we may have much clearer picture of that.

The ideal architecture in our case would be:
- View is responsible for display part (exposed to qml)
- each View has own Model (model contains data displayed in qml via View (qml must not directly call model's methods))
- ViewController manages View
- ViewController communicates with BackendController (controllers from status folder)
- BackendController is responsible for logic, async jobs, communicates with remote APIs and notifies ViewController about the changes

I guess that with this approach we may avoid redundant code, have single responsibility principe for each part and easier maintain/extend the app in the future (and most likely avoid unintentional memory leaks).

Of course if you are not ok with this, just let me know and I will easily redo everything in a way we already doing that.

**Comment by jrainville** [comment]
That's perfect. That's actually what @iurimatias 's refactor started to do. We indeed want the view to only have the display, and then the `status/` folder is what handles the backend calls, also as an abstraction to call the status-go function, in libstatus.

Great job

File: src/app/chat/views/messages.nim (hunk 1 above)
Comment (saledjenic) [reply src/app/chat/views/messages.nim:506]: Yes, I fully agree, there are multiple places where we call libstatus procedures directly from the view. I will update all such calls in messages.nim, won't go through the whole app as this is not refactoring (but every time we are fixing/developing something is an opportunity for some kind of mini refactoring).

Comment (jrainville) [reply src/app/chat/views/messages.nim:506]: Good idea!

**Comment by iurimatias** [comment]
link to status-go PR https://github.com/status-im/status-go/pull/2282

**Comment by iurimatias** [comment]
@saledjenic for an existing account I currently get `ERROR[08-02|14:58:16.506] failed to initialize db                  package=status-go/api.GethStatusBackend err="file does not exist`, would it be possible for you to rebase both status-go and this branch please.

**Pull request merged**